ext/cl_khr_integer_dot_product.asciidoc

```diff
@@ -0,0 +1,212 @@
+
+|====
+
+--
+OpenCL 3 devices must report the following feature macros via
+`CL_DEVICE_OPENCL_C_FEATURES` when the corresponding bit is set in the bitfield
+returned for `CL_DEVICE_INTEGER_DOT_PRODUCT_CAPABILITIES_KHR`:
+
+[cols="1,1",options="header"]
+|====
+| Feature bit
+| Feature macro
+
+| `CL_DEVICE_INTEGER_DOT_PRODUCT_INPUT_4x8BIT_KHR`
+| `__opencl_c_integer_dot_product_input_4x8bit`
+
+|====
+--
+
+--
+--
+
+=== Modifications to the OpenCL C Specification
+
+(Modify section 6.13.3, *Integer Functions*) ::
++
+--
+
+The following built-in functions and preprocessor definitions are added:
+
+[source,c]
+----
+#define cl_khr_integer_dot_product 1
+
+uint dot_4x8packed_uu_uint(uint a, uint b);
+int dot_4x8packed_ss_int(uint a, uint b);
+int dot_4x8packed_us_int(uint a, uint b);
+int dot_4x8packed_su_int(uint a, uint b);
+
+uint dot_acc_sat_4x8packed_uu_uint(uint a, uint b, uint acc);
+int dot_acc_sat_4x8packed_ss_int(uint a, uint b, int acc);
+int dot_acc_sat_4x8packed_us_int(uint a, uint b, int acc);
+int dot_acc_sat_4x8packed_su_int(uint a, uint b, int acc);
+
+if (CL_DEVICE_INTEGER_DOT_PRODUCT_INPUT_4x8BIT_KHR) {
+    #define __opencl_c_integer_dot_product_input_4x8bit 1
+
+    uint dot(uchar4 a, uchar4 b);
+    int dot(char4 a, char4 b);
+    int dot(uchar4 a, char4 b);
+    int dot(char4 a, uchar4 b);
+
+    uint dot_acc_sat(uchar4 a, uchar4 b, uint acc);
+    int dot_acc_sat(char4 a, char4 b, int acc);
+    int dot_acc_sat(uchar4 a, char4 b, int acc);
+    int dot_acc_sat(char4 a, uchar4 b, int acc);
+}
+----
+
+
+* `dot` returns the dot product of the two input vectors `a` and `b`. The
+components of `a` and `b` are sign- or zero-extended to the width of the
+destination type and the vectors with extended components are multiplied
+component-wise. All the components of the resulting vectors are added
+together to form the final result.
+
+* `dot_acc_sat` returns the saturating addition of the dot product of the two
+input vectors `a` and `b` and the accumulator `acc`:
+
+----
+product = dot(a,b);
+result = add_sat(product, acc);
+----
+
+* `dot_*_4x8packed_XY_R` returns the dot product of the two vectors packed
+into `a` and `b` (lowest component in least significant byte). The components
+are unpacked, sign- or zero-extended to the width of the destination type before
+the multiplications and additions. `X` represents the signedness of the components
+of `a`, `Y` that of the components of `b`. `R` is the return type.
+--
+
+=== Modifications to the OpenCL SPIR-V Environment Specification
+
+(Modify Section 5.2, *Full and Embedded Profile Extensions*) ::
++
+--
+If the OpenCL environment supports the extension `cl_khr_integer_dot_product`,
+then the environment accept modules that require `SPV_KHR_integer_dot_product` and
+declare the following SPIR-V capabilities:
+
+* *DotProductKHR*
+* *DotProductInput4x8BitKHR* if `CL_DEVICE_INTEGER_DOT_PRODUCT_INPUT_4x8BIT_KHR` is supported
+* *DotProductInput4x8BitPackedKHR*
+--
```

Comment:
One more comment (since I'm fixing #607, which is similar): this section should be moved to the SPIR-V environment spec rather than being documented here.

Reply:
Yeah, I might as well do that here. We need better cross-linking of the two documents.